autorest/codegen/__init__.py

```diff
@@ -135,12 +91,138 @@ def _build_package_dependency() -> Dict[str, str]:
+            raise ValueError(
+                "Can not add typed sync operation files if you are not showing operations. "
+                "If you want typed synced operation files, you have to add flag "
+                "--show-operations"
+            )
+
+        if options["basic_setup_py"] and not options["package_version"]:
+            raise ValueError("--basic-setup-py must be used with --package-version")
+
+        if options["package_mode"] and not options["package_version"]:
+            raise ValueError("--package-mode must be used with --package-version")
+
+        if not options["show_operations"] and options["combine_operation_files"]:
+            raise ValueError(
+                "Can not combine operation files if you are not showing operations. "
+                "If you want operation files, pass in flag --show-operations"
+            )
+
+        if options["package_mode"]:
+            if options["package_mode"] not in ("mgmtplane", "dataplane") and not Path(options["package_mode"]).exists():
+                raise ValueError(
+                    "--package-mode can only be 'mgmtplane' or 'dataplane' or directory which contains template files"
+                )
+
+        if options["reformat_next_link"] and options["version_tolerant"]:
+            raise ValueError(
+                "--reformat-next-link can not be true for version tolerant generations. "
+                "Please remove --reformat-next-link from your call for version tolerant generations."
+            )
+
+        credential_scopes = self._get_credential_scopes()
+        if credential_scopes and not options["credential"]:
+            raise ValueError("--credential-scopes must be used with the --add-credential flag")
+
+        # check to see if user just passes in --credential-scopes with no value
+        if not credential_scopes and self._autorestapi.get_boolean_value("credential-scopes", False):
+            raise ValueError(
+                "--credential-scopes takes a list of scopes in comma separated format. "
+                "For example: --credential-scopes=https://cognitiveservices.azure.com/.default"
+            )
+
+    @staticmethod
+    def _build_with_security_definition(credentials_info: Dict[str, Any], yaml_data: Dict[str, Any]):
+        security_yaml = yaml_data.get("security", {})
+        if security_yaml.get("authenticationRequired"):
+            # the key word is "AADToken" and "AzureKey"
+            credential_scopes = set()
+            key_header_name = ""
+            for scheme in security_yaml.get("schemes"):
+                if _AAD_TOKEN == scheme["type"]:
+                    credential_scopes.update(scheme["scopes"])
+                elif _AZURE_KEY == scheme["type"]:
+                    # only accept the last one
+                    key_header_name = scheme["headerName"]
+            if credential_scopes and not credentials_info.get(_Credential.SCOPES):
```

Comment:
this code is really hard to read and not clean. You can just switch the call and first build security, then build flags. Since we want flags to override security if both exist, you don't have to do checks anywhere for credentials_info.get, just update your credentials dict

Reply:
Is it ok to let null value of credential flags(like `credential_scopes`) to override defined value of security definition?

Reply:
it is fine to keep simple